soroban-sdk/src/crypto/poseidon2_sponge.rs

```diff
@@ -0,0 +1,100 @@
+use super::poseidon2_params::{get_mat_diag4_m_1, get_rc4, ROUNDS_F, ROUNDS_P, SBOX_D, T};
+use crate::{env::internal, unwrap::UnwrapInfallible, vec, Env, Symbol, TryIntoVal, Vec, U256};
+
+const RATE: u32 = 3;
+
+pub struct Poseidon2Sponge {
+    env: Env,
+    field: Symbol,
+    rate: u32,
+    squeezed: bool,
+    cache: Vec<U256>,
+    state: Vec<U256>,
+    m_diag: Vec<U256>,
+    rc: Vec<Vec<U256>>,
+}
+
+impl Poseidon2Sponge {
+    pub fn new(env: &Env, iv: U256, field: Symbol) -> Poseidon2Sponge {
+        // TODO: we need to set up parameters for BLS12-381
+        assert_eq!(field, soroban_sdk_macros::internal_symbol_short!("BN254"));
+        let mut state = vec![env];
+        for _ in 0..RATE {
+            state.push_back(U256::from_u32(env, 0));
+        }
+        state.push_back(iv);
+
+        // Get parameters from poseidon2_params
+        let m_diag = get_mat_diag4_m_1(env);
+        let rc = get_rc4(env);
+
+        Poseidon2Sponge {
+            env: env.clone(),
+            field,
+            rate: RATE,
+            cache: Vec::new(env),
+            state,
+            squeezed: false,
+            m_diag,
+            rc,
+        }
+    }
+
+    fn perform_duplex(&mut self) {
+        // zero-pad the cache
+        for _ in self.cache.len()..RATE {
+            self.cache.push_back(U256::from_u32(&self.env, 0));
+        }
+        // add the cache into sponge state
+        for i in 0..RATE {
+            let elem = self
+                .state
+                .get_unchecked(i)
+                .add(&self.cache.get_unchecked(i));
+            self.state.set(i, elem);
+        }
```

Comment:
Same comments here regarding vec construction requires less host fns if you use from_array.